code/data_structures/src/heaps/binomial_heap/binomial_heap.hpp

```diff
@@ -0,0 +1,609 @@
+	 *
+	 * returns
+	 * 		minimum value in heap list
+	 */
+	T getMin();
+	/*
+	 * extractMin
+	 * 		Removes the heap list for node with minimum value
+	 * 		and unions minimum node's children with heap list
+	 *
+	 * returns
+	 * 		minimum value in heap list
+	 */
+	T extractMin();
+	/*
+	 * deleteKey
+	 * 		Removes the node with the value v by decreasing
+	 * 		node value to one less than current min and then
+	 * 		extracts the min.
+	 * value:
+	 * 		value to delete from heap
+	 *
+	 * returns
+	 * 		true if successfully deleted
+	 */
+	bool deleteKey(T value);
+	/*
+	 * decreaseKey
+	 * 		Searches for node with value v a
+	 *
+	 * 	value:
+	 * 		value to find and decrease
+	 * 	newValue:
+	 * 		value to decrease value to
+	 * returns
+	 * 		true if successfully decreased
+	 */
+	bool decreaseKey(T value, T newValue);
+	/*
+	 * size
+	 * 		Number of nodes in heap.
+	 *
+	 * returns
+	 *		number of nodes in heap
+	 */
+	int size();
+	bool isEmpty();
+	void print(std::ostream& out);
+protected:
+	/*Utility heap functions*/
+	/*
+	 * findKey
+	 * 		Searches heap for node with value equal to key.
+	 * 		v : value in heap to find
+	 * 	returns
+	 * 	 a pointer to node with value key
+	 */
+	binomialNode<T>* findKey(T v);
+	/*
+	 * unionHeap
+	 * 		Consolidates two heaps by merging them into a single
+	 * 		The heap is appended to the end of the child list to maintain child order
+	 * 		of roots from min to max
+	 * 		Assumes heapA and heapB siblings are ordered by root order(sibling count)
+	 *
+	 * 	HeapA - root of heap to union with heapB
+	 * 	HeapB - root of heap to union with heapA
+	 *
+	 *
+	 * 	returns
+	 * 		the root of the heap representing the union of heapA and heapB
+	 * 		such that the either heapA becomes a child of heapB or the reverse
+	 */
+	static binomialNode<T>* unionHeap(binomialNode<T>* heapA, binomialNode<T>* heapB);
+	/* mergeHeap
+	 * 		Merges two heaps into a single binomial heap by adding
+	 * 		all siblings of heapA and heapB into a single heap containing
+	 * 		all nodes in heapA and heapB.
+	 * 		Assumes heapA and heapB siblings are ordered by root order(sibling count)
+	 * 		Consolidating two heaps maintains merged ordering
+	 *
+	 * 	HeapA - root of heap to union with heapB
+	 * 	HeapB - root of heap to union with heapA
+	 *
+	 * 	returns
+	 * 		the root of the new heap that contains all heapA and heapB nodes
+	 *
+	 */
+	static binomialNode<T>* mergeHeap(binomialNode<T>* heapA, binomialNode<T>* heapB);
+	/*
+	 * order
+	 * 		Number of children incident on heap
+	 * 	heap
+	 * 		pointer to root of binomial heap
+	 * 	returns
+	 * 		order or heap
+	*/
+	static int order(binomialNode<T>* heap);
+private:
+	/*
+	 * head
+	 * 		pointer to the head of the heap list where every
+	 * 		node is the root node of a heap
+	 */
+	binomialNode<T> * head;
+};
+
+
+template <typename T>
+binomialHeap<T>::binomialHeap()
+{
+	head = 0;
+}
+
+template <typename T>
+binomialHeap<T>::~binomialHeap()
+{
+	if (head)
```

Comment:
No need for the check. Make this destructor virtual. (Since there are possibilities for a derived class)